dlt/destinations/impl/lancedb/lancedb_client.py

```diff
@@ -696,67 +694,149 @@ def complete_load(self, load_id: str) -> None:
-                # We add and set these missing fields, that are fields not present in the target schema, to NULL.
-                missing_fields = set(self._load_table["columns"]) - set(record)
-                for field in missing_fields:
-                    record[field] = None
+        # We need upsert merge's deterministic _dlt_id to perform orphan removal.
+        # Hence, we require at least a primary key on the root table if the merge disposition is chosen.
+        if (
+            (self._load_table not in self._schema.dlt_table_names())
+            and not self._load_table.get("parent")  # Is root table.
+            and (write_disposition == "merge")
+            and (not get_columns_names_with_prop(self._load_table, "primary_key"))
+        ):
+            raise DestinationTerminalException(
+                "LanceDB's write disposition requires at least one explicit primary key."
+            )
 
-        upload_batch(
-            records,
-            db_client=self._db_client,
-            table_name=self._fq_table_name,
+        write_records(
+            arrow_table,
+            db_client=db_client,
+            table_name=fq_table_name,
             write_disposition=write_disposition,
-            id_field_name=self._id_field_name,
+            merge_key=self._schema.data_item_normalizer.C_DLT_ID,  # type: ignore[attr-defined]
         )
+
+
+class LanceDBRemoveOrphansJob(RunnableLoadJob):
+    orphaned_ids: Set[str]
+
+    def __init__(
+        self,
+        file_path: str,
+    ) -> None:
+        super().__init__(file_path)
+        self._job_client: "LanceDBClient" = None
+        self.references = ReferenceFollowupJobRequest.resolve_references(file_path)
+
+    def run(self) -> None:
+        dlt_load_id = self._schema.data_item_normalizer.C_DLT_LOAD_ID  # type: ignore[attr-defined]
+        dlt_id = self._schema.data_item_normalizer.C_DLT_ID  # type: ignore[attr-defined]
+        dlt_root_id = self._schema.data_item_normalizer.C_DLT_ROOT_ID  # type: ignore[attr-defined]
+
+        db_client: DBConnection = self._job_client.db_client
+        table_lineage: TTableLineage = [
+            TableJob(
+                table_schema=self._schema.get_table(
+                    ParsedLoadJobFileName.parse(file_path_).table_name
+                ),
+                table_name=ParsedLoadJobFileName.parse(file_path_).table_name,
+                file_path=file_path_,
+            )
+            for file_path_ in self.references
+        ]
+
+        for job in table_lineage:
+            target_is_root_table = "parent" not in job.table_schema
+            fq_table_name = self._job_client.make_qualified_table_name(job.table_name)
+            file_path = job.file_path
+            with FileStorage.open_zipsafe_ro(file_path, mode="rb") as f:
+                payload_arrow_table: pa.Table = pq.read_table(f)
+
+            if target_is_root_table:
```

Comment:
OK this is really good but IMO there's a mistake when deleting nested tables:
- to delete orphan from root table we use merge key of the root table (good)
- to delete from child table we use all `dlt_ids` that you can find in root table - not in child table why? because if you remove (1) all nested elements in root table you will not delete all orphans (2) or you remove one row from root table you also won't find the proper root key

please write tests for that.

also could you take taxi dataset and try to merge all rows (where IN clause will be insane long). I wonder what happens: will that be o(n^2) operation or they will do a hash table on IN clause

Reply:
Setting a scalar index does seem to offers constant time complexity. Clearly this is worth investigating for future PRs.